build: Build Swift SDK for WebAssembly

kateinoigakukun opened the pull request:
This change adds a new product, WasmSwiftSDK, to build the Swift SDK for WebAssembly. The product is built using the swift-sdk-generator package and takes just built WebAssembly stdlib, clang runtime libraries, and wasi-sysroot as input, and produces a Swift SDK artifactbundle under swift-sdk-generator/Bundles.

The main purpose of building SDK on CI is to test WebAssembly target-specific code in SwiftPM like test runner for now

kateinoigakukun committed:
build: Build Swift SDK for WebAssembly

This change adds a new product, WasmSwiftSDK, to build the Swift SDK for
WebAssembly. The product is built using the swift-sdk-generator package
and takes just built WebAssembly stdlib, clang runtime libraries,
and wasi-sysroot as input, and produces a Swift SDK artifactbundle under
swift-sdk-generator/Bundles.

kateinoigakukun commented:
@swift-ci test WebAssembly platform

kateinoigakukun commented:
Hmm, CI is still using Swift 5.8 but swift-sdk-generator requires 5.9. We might need to wait https://github.com/apple/swift-docker/pull/368 or relax minimum version requirement on swift-sdk-generator? CC: @MaxDesiatov

MaxDesiatov commented:
I think lowering the requirement to 5.8 should be possible, I will look into that.

finagolfin commented on utils/swift_build_support/swift_build_support/products/wasmstdlib.py:
```diff
@@ -194,3 +197,94 @@ def add_extra_cmake_options(self):
+                target_triple, stdlib_build_path, llvm_runtime_libs_build_path)
+            target_packages.append((target_triple, package_path))
+
+        swiftc_path = os.path.abspath(self.toolchain.swiftc)
```
Why are you using the prebuilt 5.8 toolchain on the CI after building a fresh trunk 6.1 toolchain? Most build products use [that freshly built trunk `native_toolchain_path()`](https://github.com/apple/swift/blob/ae6ff02d1be440faa6e4e14296b04fbb02d341fb/utils/swift_build_support/swift_build_support/products/swiftpm.py#L49).

kateinoigakukun replied:
The Swift SDK will be used to test SwiftPM, so it should be built before SwiftPM and we cannot use just built SwiftPM here.

finagolfin replied:
OK, you are currently building this Wasm SDK at the end though, after a bunch of SwiftPM products are also built with trunk SwiftPM. You will have to move the Wasm SDK build much earlier to use it to test SwiftPM.

kateinoigakukun replied:
Right, the current position in the pipeline is too late.

MaxDesiatov commented:
@swift-ci test WebAssembly

kateinoigakukun commented:
@swift-ci test WebAssembly

MaxDesiatov commented:
@swift-ci test WebAssembly

MaxDesiatov commented:
Looks like there's some bug with CI status propagation. A build seems to be running, but its status is not reflected in the list of current PR checks 😞

MaxDesiatov commented on utils/swift_build_support/swift_build_support/products/wasmstdlib.py:
```diff
@@ -194,3 +197,94 @@ def add_extra_cmake_options(self):
         self.cmake_options.define('SWIFT_ENABLE_WASI_THREADS:BOOL', 'TRUE')
+
+
+class WasmSwiftSDK(product.Product):
```
Since this is WASI-specific, would it make sense to reflect that in the product name?

```suggestion
class WASISwiftSDK(product.Product):
```

kateinoigakukun replied:
This SDK would eventually support multiple targets within the single SDK, so it's not WASI spesific.

kateinoigakukun committed:
update-checkout: Update swift-sdk-generator dependencies

swift-nio and swift-async-algorithms have been updated to support
`SWIFTCI_USE_LOCAL_DEPS` convention. This change updates them to the
versions including the support.

kateinoigakukun committed:
build: Build Swift SDK earlier than SwiftPM to allow SwiftPM to use it

Just built Swift SDK would be useful for SwiftPM to perform integration
tests. All of those products do not depend on just built SwiftPM
(WasmKit is built using host tools), so it is safe to build them earlier.

kateinoigakukun commented:
@swift-ci test WebAssembly

kateinoigakukun commented:
@swift-ci test WebAssembly

kateinoigakukun commented:
@swift-ci test WebAssembly

kateinoigakukun commented:
@swift-ci test WebAssembly

kateinoigakukun commented:
@swift-ci smoke test

edymtt commented on utils/swift_build_support/swift_build_support/products/wasmstdlib.py:
```diff
@@ -206,3 +209,101 @@ def add_extra_cmake_options(self):
         self.cmake_options.define('SWIFT_ENABLE_WASI_THREADS:BOOL', 'TRUE')
+
+
+class WasmSwiftSDK(product.Product):
+    @classmethod
+    def product_source_name(cls):
+        return "swift-sdk-generator"
```
It could make more sense to move the entire `WasmSwiftSDK` class in a separate file (like we do for almost all build products except `WASILibc` and `WasmLLVMRuntimeLibs`) -- is there any reason for which we put this in the same file as `WasmStdlib`?

kateinoigakukun replied:
That makes sense to me 👍

kateinoigakukun commented:
@swift-ci smoke test

kateinoigakukun commented:
@swift-ci test WebAssembly

kateinoigakukun commented:
@swift-ci test WebAssembly

kateinoigakukun commented:
@swift-ci smoke test

kateinoigakukun commented:
@swift-ci test WebAssembly

kateinoigakukun commented:
@swift-ci smoke test Linux platform

kateinoigakukun commented:
@swift-ci smoke test

kateinoigakukun commented:
@shahmishal Could you add swift-sdk-generator repository to checkout in CI infrastructure? :pray:

shahmishal commented on utils/update_checkout/update-checkout-config.json:
```diff
@@ -123,7 +125,7 @@
                 "swift-certificates": "1.0.1",
                 "swift-asn1": "1.0.0",
-                "swift-async-algorithms": "1.0.0",
+                "swift-async-algorithms": "main",
```
Is there a reason we are using main branch now?

kateinoigakukun replied:
We need to use a revision including https://github.com/apple/swift-async-algorithms/commit/6ae9a051f76b81cc668305ceed5b0e0a7fd93d20 to make SDK generator build work.
But it's not ideal situation, so I asked a new release including the fix just now https://github.com/apple/swift-async-algorithms/issues/314

kateinoigakukun commented:
@shahmishal Gentle ping🙏
